koswat/strategies/infra_priority_strategy/infra_priority_strategy.py

```diff
@@ -0,0 +1,108 @@
+from collections import defaultdict
+from itertools import groupby
+
+from koswat.dike_reinforcements.reinforcement_profile.reinforcement_profile_protocol import (
+    ReinforcementProfileProtocol,
+)
+from koswat.strategies.order_strategy.order_strategy import OrderStrategy
+from koswat.strategies.strategy_input import StrategyInput
+from koswat.strategies.strategy_location_reinforcement import (
+    StrategyLocationReinforcement,
+)
+from koswat.strategies.strategy_protocol import StrategyProtocol
+
+
+class InfraPriorityStrategy(StrategyProtocol):
+    """
+    This strategy works under the assumption that
+    `StrategyLocationReinforcement.available_measures`
+    ARE NOT ordered by "compatibility".
+    Thus we need to check first which are the compatible reinforcements
+    per cluster.
+    """
+
+    def _get_initial_clusters(
+        self, strategy_input: StrategyInput
+    ) -> list[StrategyLocationReinforcement]:
+        return OrderStrategy().apply_strategy(strategy_input)
+
+    def _get_common_available_measures(
+        self,
+        initial_measure: type[ReinforcementProfileProtocol],
+        clustered_locations: list[StrategyLocationReinforcement],
+    ) -> dict[type[ReinforcementProfileProtocol], float]:
+        # Define initial costs
+        _costs_dict = defaultdict(lambda: 0.0)
+        _costs_dict[initial_measure] = sum(
+            _c.current_cost for _c in clustered_locations
+        )
+
+        # Get common measures excluding the initial one to reduce computations.
+        _common_available_measures = set(
+            OrderStrategy.get_default_order_for_reinforcements()
+        )
+        _common_available_measures.discard(initial_measure)
+
+        # Intersect per location
+        for _cl in clustered_locations:
+            _common_available_measures.intersection_update(_cl.available_measures)
+            _discarded_measures = set()
+            # Get only available measures **cheaper** than current state  and their cost.
+            for _am in _common_available_measures:
+                _costs_dict[_am] += clustered_locations[0].get_reinforcement_costs(_am)
```

Comment:
Would it make sense to get the cost per reinforcement prior to these loops to save "ticks"?

Reply:
That is already what I'm trying to do here, as the reinforcement costs here include also the location ones (and that's why it's a summation ( `+=`) instead of an assignment

Reply:
Yes, I see. However, reinforcement cost are constant per location, so you could just get the first location and multiply by the number of locations before the loop already.